aptos-move/aptos-aggregator/src/transaction.rs

```diff
@@ -40,9 +42,111 @@ impl ChangeSetExt {
+            if let Some(r) = write_ops.get_mut(&key) {
+                match r {
+                    Creation(data) => {
+                        let val: u128 = bcs::from_bytes(data)?;
+                        *r = Creation(bcs::to_bytes(&op.apply_to(val)?)?);
+                    }
+                    Modification(data) => {
+                        let val: u128 = bcs::from_bytes(data)?;
+                        *r = Modification(bcs::to_bytes(&op.apply_to(val)?)?);
+                    }
+                    Deletion => {
+                        bail!("Failed to apply Aggregator delta -- value already deleted");
+                    }
+                }
+            } else {
+                match delta_ops.entry(key) {
+                    Occupied(entry) => {
+                        entry.into_mut().merge_with(op)?;
+                    }
+                    Vacant(entry) => {
+                        entry.insert(op);
+                    }
+                }
+            }
+        }
+
+        Ok(Self {
+            delta_change_set: delta,
+            change_set: ChangeSet::new(write_set.freeze()?, events),
+        })
+    }
+
+    pub fn squash_change_set(self, other: ChangeSet) -> anyhow::Result<Self> {
+        use btree_map::Entry::*;
+        use WriteOp::*;
+
+        let (mut delta, change_set) = self.into_inner();
+        let (write_set, mut events) = change_set.into_inner();
+        let mut write_set = write_set.into_mut();
+        let write_ops = write_set.as_inner_mut();
+
+        let (other_write_set, other_events) = other.into_inner();
+
+        for (key, op) in other_write_set.into_iter() {
+            match write_ops.entry(key) {
+                Occupied(mut entry) => {
+                    let r = entry.get_mut();
+                    match (&r, op) {
+                        (Modification(_) | Creation(_), Creation(_))
+                        | (Deletion, Deletion | Modification(_)) => {
+                            bail!("The given change sets cannot be squashed")
+                        }
+                        (Modification(_), Modification(data)) => *r = Modification(data),
+                        (Creation(_), Modification(data)) => *r = Creation(data),
+                        (Modification(_), Deletion) => *r = Deletion,
+                        (Deletion, Creation(data)) => *r = Modification(data),
+                        (Creation(_), Deletion) => {
+                            entry.remove();
+                        }
+                    }
+                }
+                Vacant(entry) => {
+                    delta.remove(entry.key());
+                    entry.insert(op);
+                }
+            }
+        }
+
+        events.extend(other_events);
+
+        Ok(Self {
+            delta_change_set: delta,
+            change_set: ChangeSet::new(write_set.freeze()?, events),
+        })
+    }
+
+    pub fn squash(self, other: Self) -> anyhow::Result<Self> {
```

Comment:
@gelash @georgemitenkov I've implemented the correct (hopefully) algorithm. Please have a look and let me know if it looks good to you.

Reply:
Okay I gotta land this to unblock myself for the cut, but do let me know if you spot anything wrong. I'd be happy to fix them at a later time.

Reply:
makes sense, and LGTM, let's add tests and make generally usable at a later time.
Did this allow passing the e2e or was it smt else?

Reply:
It was something else -- more specifically, I had to change from returning an error to panicking when failing to materialize the delta change set to match the old behavior. We may want to revisit this though.